initialization/initialization_test.go

```diff
@@ -41,425 +46,466 @@ func (l testLogger) Debug(msg string, args ...interface{}) { l.t.Logf("\tDEBUG\t
+		WithInitOpts(opts),
+		WithLogger(testLogger{t: t}),
+	)
 	r.NoError(err)
-	err = init.Initialize()
-	r.ErrorAs(err, cfgMissErr)
+	r.ErrorAs(init.Initialize(context.Background()), cfgMissErr)
 	r.Equal("NumUnits", cfgMissErr.Param)
-
-	// Cleanup.
-	r.NoError(init.Reset())
 }
 
 func TestInitialize_MultipleFiles(t *testing.T) {
 	r := require.New(t)
 
-	cfg, opts := getTestConfig(t)
-	commitment := make([]byte, 32)
-	init, err := NewInitializer(cfg, opts, commitment)
-	r.NoError(err)
-	err = init.Initialize()
-	r.NoError(err)
+	cfg := config.DefaultConfig()
+	cfg.LabelsPerUnit = 1 << 12
 
-	oneFileData, err := initData(opts.DataDir, uint(cfg.BitsPerLabel))
+	opts := config.DefaultInitOpts()
+	opts.DataDir = t.TempDir()
+	opts.NumUnits = cfg.MinNumUnits
+	opts.NumFiles = 2
+	opts.ComputeProviderID = CPUProviderID()
+
+	init, err := NewInitializer(
+		WithCommitment(make([]byte, 32)),
+		WithConfig(cfg),
+		WithInitOpts(opts),
+		WithLogger(testLogger{t: t}),
+	)
 	r.NoError(err)
+	r.NoError(init.Initialize(context.Background()))
 
-	// Cleanup.
-	err = init.Reset()
+	oneFileData, err := initData(opts.DataDir, uint(cfg.BitsPerLabel))
 	r.NoError(err)
 
 	for numFiles := uint32(2); numFiles <= 16; numFiles <<= 1 {
-		opts := opts
-		opts.NumFiles = numFiles
-
-		init, err := NewInitializer(cfg, opts, commitment)
-		r.NoError(err)
-		err = init.Initialize()
-		r.NoError(err)
-
-		multipleFilesData, err := initData(opts.DataDir, uint(cfg.BitsPerLabel))
-		r.NoError(err)
-
-		r.Equal(multipleFilesData, oneFileData)
-
-		// Cleanup.
-		r.NoError(init.Reset())
+		t.Run(fmt.Sprintf("NumFiles=%d", numFiles), func(t *testing.T) {
+			opts := opts
+			opts.NumFiles = numFiles
+			opts.DataDir = t.TempDir()
+
+			init, err := NewInitializer(
+				WithCommitment(make([]byte, 32)),
+				WithConfig(cfg),
+				WithInitOpts(opts),
+				WithLogger(testLogger{t: t}),
+			)
+			r.NoError(err)
+			r.NoError(init.Initialize(context.Background()))
+
+			multipleFilesData, err := initData(opts.DataDir, uint(cfg.BitsPerLabel))
+			r.NoError(err)
+
+			r.Equal(multipleFilesData, oneFileData)
+		})
 	}
 }
 
 func TestNumLabelsWritten(t *testing.T) {
-	req := require.New(t)
+	r := require.New(t)
 
-	cfg, opts := getTestConfig(t)
-	commitment := make([]byte, 32)
-	init, err := NewInitializer(cfg, opts, commitment)
-	req.NoError(err)
+	opts := getTestOptions(t)
+	init, err := NewInitializer(opts...)
+	r.NoError(err)
 
 	// Check initial state.
 	numLabelsWritten, err := init.diskState.NumLabelsWritten()
-	req.NoError(err)
-	req.Equal(uint64(0), numLabelsWritten)
+	r.NoError(err)
+	r.Equal(uint64(0), numLabelsWritten)
 
 	// Initialize.
-	err = init.Initialize()
-	req.NoError(err)
+	r.NoError(init.Initialize(context.Background()))
 	numLabelsWritten, err = init.diskState.NumLabelsWritten()
-	req.NoError(err)
-	req.Equal(uint64(opts.NumUnits)*cfg.LabelsPerUnit, numLabelsWritten)
+	r.NoError(err)
+	r.Equal(uint64(init.opts.NumUnits)*init.cfg.LabelsPerUnit, numLabelsWritten)
```

Comment:
IMO, the test should not access the private fields of the tested thing. Why not check against local `opts`?

Reply:
Depends; in a unit test testing only the Initializer I think it is OK to access private fields. I'm not testing behaviour here, I'm just accessing data. It's not unusual to do this in go (example test from the standard library for `context` where a test checks that a parent context actually contains their children: https://cs.opensource.google/go/go/+/refs/tags/go1.19.3:src/context/context_test.go;l=158-167)

Testing this only via public methods would either require getter for these values (which we don't want to add just because of a test) or make the test more complicated.

The specific reason I'm doing it here is so that I don't need to re-define `opts` and `cfg` for this test. Both of those are hidden behind `opts := getTestOptions(t)` which returns a slice of functions that I would need to execute to get access to those values. Alternatively I could inline `getTestOptions` (it is only used here and in one other test); if that is a compromise you are OK with?

If I would test the interaction with another component I would completely agree with you.

Reply:
There are two issues with unit tests accessing private fields.

1. They assume certain implementation. It impedes refactoring.
2. The test uses private data of `init` as the expected value in the assertion. It assumes that `init.opts` is properly assigned in `NewInitializer(opts...)`. If it is not properly assigned because of a bug, then the assertion would be invalid.

I think that inlining `getTestOptions` as you suggested is the way to go 😉 .